### ISSUE TITLE
Tracking Issue for `Exclusive`

### ISSUE THREAD

**Issue by guswynn** [rust-lang/rust#98407]
Feature gate: `#![feature(exclusive_wrapper)]`

This is a tracking issue for the `Exclusive` wrapper struct.

This structure can be used to upgrade non-`Sync` objects into `Sync` ones, by only offering mutable access to the underlying object. This can be useful with `Send`-but-not-`Sync` futures (like `BoxFuture`s)

### Public API

```rust
// core::sync
#[derive(Default)]
struct Exclusive<T: ?Sized> { ... }

impl<T: ?Sized> Sync for Exclusive {}

// via auto traits:
// impl<T> Send for Exclusive<T>
// where
//    T: Send + ?Sized,

impl<T> Exclusive<T> {
    pub const fn new(t: T) -> Self;
    pub const fn into_inner(self) -> T;
}

impl<T: ?Sized> Exclusive<T> {
    pub const fn get_mut(&mut self) -> &mut T;
    pub const fn get_pin_mut(Pin<&mut self>) -> Pin<&mut T>;
    pub const fn from_mut(&mut T) -> &mut Exclusive<T>;
    pub const fn from_pin_mut(Pin<&mut T>) -> Pin<&mut Exclusive<T>>;
}

impl<T: Future> Future for Exclusive { ... }

impl<T> From<T> for Exclusive<T> { ... }
impl<T: ?Sized> Debug for Exclusive { ... }
```

### Steps / History

- [x] Implementation: https://github.com/rust-lang/rust/pull/97629
- [x] Determine other nice-to-have trait impls
  - [x] https://github.com/rust-lang/rust/pull/116241
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

**Comment by clarfonthey** [comment]
This is probably a bikeshed, but perhaps the `get_pin_mut` method could be directly implemented on `Pin<Exclusive<T>>`, so that it can share the same name as the regular `Exclusive<T>` version? Then maybe both could be named `get_mut`, or even `get` since you'll have to use it to access non-mutable methods, just with exclusive access.

**Comment by guswynn** [comment]
@clarfonthey that would conflict with https://doc.rust-lang.org/std/pin/struct.Pin.html#method.get_mut

**Comment by Nemo157** [comment]
To provide some motivation I have been using a more limited internal version of this for about 3 years [in `async-compression`](https://docs.rs/crate/async-compression/latest/source/src/unshared.rs). This is the sole bit of unsafety in that crate so I would love to be able to offload that onto `std`.

**Comment by guswynn** [comment]
@Nemo157 I think we should try to get this in at 1.66, next week ill take a look at writing a stabilization report

**Comment by H4x5** [comment]
Hi! I've submitted PR #104057, with a few small changes:
- Added `rustc_const_unstable` attributes on the const fns.
  - `into_inner` is gated under `const_exclusive_into_inner`, and is blocked on const drop.
  - `{get,from}_{mut,pin_mut}` is gated under `const_exclusive_get_mut`, and is blocked on `const_mut_refs`.
- The From impl is now unstably const under `const_convert`, for consistency with the rest of the standard library.
- `Exclusive<G>` now implements Generator, gated under `generator_trait`.
- `Exclusive<F>` now implements `FnOnce` & `FnMut`.

I would be happy to write a stabilization PR when this passes an FCP.

**Comment by zachs18** [comment]
For the same reason that `Exclusive` can unconditionally implement `Sync`, I believe it can also unconditionally implement [`core::panic::RefUnwindSafe`](https://doc.rust-lang.org/core/panic/trait.RefUnwindSafe.html) (which claims that a shared reference to the type is unwind-safe (i.e. can cross a panic boundary without introducing logic errors), which `&Exclusive` is, because it has no API).

I don't know if it would be that useful (even ignoring that there's been [some disucssion](https://github.com/rust-lang/rfcs/pull/3260) about deprecating `(Ref)UnwindSafe`), but I believe it could be added after stabilization if it is wanted, so it's not urgent or blocking.

**Comment by harudagondi** [comment]
Making the inner type public would make it accessible with shared references, which defeats the point of `Exclusive` to begin with.

**Comment by harudagondi** [comment]
Interestingly, in bevyengine/bevy#7718, they are adding a method called `read` that allows shared access to their `Exclusive` analogue as long as the inner type implements `Sync`.

**Comment by alice-i-cecile** [comment]
> Q: What's the use case for a SyncCell when the type is already Sync?
>  A: Generics, mostly. When designing an API you might not know if a type parameter is Sync or not, so you'd put it in a SyncCell. Then when you use that API with a concrete type, you can bypass the exclusivity if that type is Sync

The method above is quite useful for us, but is a contradiction with the `Exclusive` name of this type. As a result, we'd like to suggest renaming the type to something compatible, and adding the method linked in the PR above.

This is quite a useful type, and it would be a shame to see something limited get locked in via stabilization.

(summarized from [conversation in the Bevy Discord](https://discord.com/channels/691052431525675048/749335865876021248/1076198832460795915)).

**Comment by zachs18** [comment]
I agree that `SyncCell` or something like it could be a better name than `Exclusive` in the prescence of a `fn read(&self) -> &T where T: Sync`-like method, but I don't necessarily think that `Exclusive` being a contradictory name/having an exception for `Exclusive<T: Sync>` would be too bad. There's somewhat of precendent with `Pin<Pointer>`, which has an "exception" when `Pointer::Target: Unpin`, and indeed has additional API in that case, e.g. [`Pin::get_mut`](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.get_mut).

**Comment by JoJoJet** [comment]
I don't think `Pin` is a the best comparison here. It makes total sense that the `Unpin` trait would negate `Pin`, for obvious reasons. However, it does not make sense that `Sync` would negate `Exclusive`, since those concepts are orthogonal. If shared access is going to be supported for this type, I don't think it makes sense to keep the current name.

**Comment by clarfonthey** [comment]
The main issue with using the `Cell` terminology I can see is that `Exclusive` doesn't use `UnsafeCell` internally, whereas all other forms of `Cell` do. In some sense, this actually makes it a sort of anti-cell; cells let you perform writes with immutable references, whereas `Exclusive` prevents you from performing reads with immutable references.

What we have is some sort of static mutex, but I'm not sure if using that terminology is helpful either.

**Comment by JoJoJet** [comment]
I agree that the `Cell` terminology isn't great, since this type doesn't use interior mutability.

I think the name should emphasize the fact that the purpose of this type is to gurantee `Sync`-ness. A name such as `SyncSafe<T>` would be ideal IMO -- this describes the purpose of this type, and doesn't make shared access contradictory.

**Comment by Nikita240** [comment]
We should just name it `Iceberg` since the type is used to `Sync` the `Unsyncable`.

**Comment by Xuanwo** [comment]
Is it possible (and how) to make `Exclusive` implement traits that only have `&mut self` APIs? There are two useful cases:

- Trait in `std` or 3rd crates, like `Read` and `Write`.
- Trait in users own crates, like `MyMutOnlyTrait`.

**Comment by zachs18** [comment]
> You could also "cheat" by implementing traits that have read-only methods specifically for `&mut Exclusive`, like for example `Iterator` could be done with this approach.

I don't think that would work? If you have a `&self` method, where `Self = &mut Exclusive`, then you have `&&mut Exclusive`, which you cannot get a `&mut Exclusive` from.

Well, `Iterator` specifically could technically be implemented, since the only `&self` method `size_hint` can be implemented conservatively as `(0, None)` without accessing the iterator, but in general this is not true.


<details> <summary>Trying to implement <Exclusive as Iterator>::size_hint normally</summary>

([Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=81e301b24070ed88ad3778875a3bdc6a))

```Rust
impl<I: ?Sized + Iterator> Iterator for &mut Exclusive<I> {
    type Item = I::Item;
    
    fn next(&mut self) -> Option<I::Item> {
        self.get_mut().next() // this is fine
    }
    
    fn size_hint(&self) -> (usize, Option<usize>) {
        self.get_mut().size_hint()
        // cannot borrow `**self` as mutable, as it is behind a `&` reference
    }
}
```

</details>

**Comment by clarfonthey** [comment]
…right, initially I thought that reborrows of the initial `&mut` would not violate the exclusive access guarantees, but in hindsight, they do. You're right that we'd have to rely on naïve implementations of `size_hint`.

**Comment by GoldsteinE** [comment]
Is there anything blocking this from stabilization? It looks old enough at this point and I don’t see any open concerns in this thread. It looks like it just waits for an FCP?

**Comment by Scripter17** [comment]
Am I allowed to put a bounty on this? If so, $10 USD to whoever starts the FCP that ends in a merge (paid once the bot says this is merged since I don't think there's anything that'll cancel stabilization after that (if it does happen you don't need to refund me and for the lols I'll also add a bounty on the second merging FCP)).

If I'm not allowed to put a bounty on this, sorry for the notification and also the next one where someone tells me I'm not allowed to put a bounty on this.

I prefer paypal but I can probably work with other stuff. If there's sanctions in the way I can probably mail you something worth $10.

**Comment by clarfonthey** [comment]
FCPs rely pretty much entirely on the bandwidth of the team approving the change, not the person writing the stabilisation PR.

**Comment by Scripter17** [comment]
> FCPs rely pretty much entirely on the bandwidth of the team approving the change, not the person writing the stabilisation PR.

As much as I love setting bounties I don't have the margin to do $10 for every merge vote.

Just getting someone to start the FCP basically guarantees a merge soon since it's <200 lines of a fairly straightforward feature. I realize even that is worth at least $20 per person but even just $10 is a pretty big chunk of my bounty money.

**Comment by clarfonthey** [comment]
I guess my point is that it seems unlikely that, even if someone were to accept the bribe, it would actually perceptibly merge faster. A stabilisation PR and an FCP take no time at all to write for a small change like this and I'm not sure if people would feel comfortable prioritising this feature over others in exchange of money.

**Comment by Scripter17** [comment]
In light of the rapid consensus, I'll be sure to not do that again. Sorry.

**Comment by 197g** [comment]
Let's start with the formal part of process, a proposed stabilization report:

---

# Stabilization Report

## Implementation History

- The type was in #97629 with its constructions and interation with `Pin`/`Future`
- In #116241 a few further traits on `&mut` were identified, `FnMut`, `FnOnce` and the unstable `Generator` to expand on `Future`.

## Api summary

See [the tracking issue top comment](https://github.com/rust-lang/rust/issues/98407#issue-1280866325).

## Experience report

In bevy, a widely used entity component system, [`SyncCell`](https://docs.rs/bevy_utils/latest/bevy_utils/synccell/struct.SyncCell.html) is provided in its `bevy-utils` crate and was always intended as a temporary solution to [`Exclusive`'s initial design](https://github.com/bevyengine/bevy/pull/5483#issuecomment-1234805001). It facilitates wrapping existing types. All `Resources` managed by the ECS in the design must be `Sync`, yet a pass over _mutable references_ to a type of resource may want access to attributes of the resource that are not `Sync` themselves. See [bevy's documentation as well](https://docs.rs/bevy/latest/bevy/prelude/trait.Resource.html#sync-resources). `Exclusive` can wrap such fields to fulfill both requirements simultaneously.

`bevy` extended the API with one additional shared getter method :

```
pub fn read(&self) -> &T
where
    T: Sync,
```

In some situations, a generic struct may choose to unconditionally wrap a field but a concrete instantiation may be able to prove the field to be `Sync` to undo the extra invariants imposed. This is a forward compatibly case that can be readily addressed and stabilized separately.

This same design is also seen in other crates, e.g. https://github.com/clstatham/weaver/blob/491488e33c1e8a493a23885bb06470bde40e2819/crates/weaver-util/src/lock.rs#L247-L276

---

**Comment by Jules-Bertholet** [comment]
Some concerns:

- Should there be an `impl<T: ?Sized> AsMut<T> for Exclusive<T>`? (This would be a breaking change to add later)
- Should `Exclusive`’s methods be changed to associated functions, in order to leave room for a potential future language extension that would allow `Exclusive` to implement `DerefMut` without `Deref`? (It would not be a breaking change to go from associated functions to methods, if we later decide this definitely won’t happen)

**Comment by Jules-Bertholet** [comment]
> Genuine question: How could Rust and `DerefMut` be redesigned to allow for that?

https://github.com/rust-lang/rfcs/pull/3437

> I assume they'd be unrelated?

Most likely, they would both have `Receiver` (new trait being added for arbitrary self types) as a supertrait.

**Comment by RustyYato** [comment]
> > Genuine question: How could Rust and DerefMut be redesigned to allow for that?
> 
> https://github.com/rust-lang/rfcs/pull/3437

I don't see how trait aliases allow you to avoid the problematic `Deref` impl. Can you elaborate on this?

I don't think this can work because I can do `fn get_deref<T: DerefMut>(t: &T) -> &T::Target { t }`, and any future Rust version cannot change this to mean something other than calling `Deref::deref`. And since `Exclusive` cannot implement `Deref`, I don't see how you think we can implement `DerefMut`, which requires implementing `Deref`.

It may be possible to special-case `Exclusive` to allow you to coerce `&mut Exclusive<T>` to `&mut T`, but that seems separate to `DerefMut`.

**Comment by Jules-Bertholet** [comment]
> I don't see how trait aliases allow you to avoid the problematic `Deref` impl. Can you elaborate on this?

We would introduce a new trait `pub trait DerefMutOnly: Receiver { fn deref_mut(&mut self) -> &mut Self::Target; }`, and redefine `DerefMut` as an alias `pub trait DerefMut = DerefMutOnly where Self: Deref;`. Existing `DerefMut` bounds would retain their existing behavior of implying `Deref`, so no existing code would break.

**Comment by Jules-Bertholet** [comment]
> We felt that `AsMut` […] did not need to be blockers for stabilization; they could always be added later

No, it could not be added later. That would be a major breaking change, as downstream crates are allowed to do this:

```rust
struct Foo;

impl AsMut<Foo> for core::sync::Exclusive<Foo> {
    fn as_mut(&mut self) -> &mut Foo {
        self.get_mut()
    }
}
```

(That being said, we have gotten away with technically-breaking changes of this sort in the past, when crater has said it’s OK. But why risk it?)

> We'd be interested in seeing a `get_ref` method added that requires `T: Sync`, like Bevy's version has, but that's also not a blocker for stabilization.

Is there a risk that adding this later could cause soundness issues (if users come to rely on it not being provided)?

Also, I suppose we could also add a `Deref` impl with a `T: Sync` bound.

**Comment by RustyYato** [comment]
I don't think we should drop `get_mut`, since that can be `const`. While const traits are unlikely to be stabilized in the near term.

**Comment by Jules-Bertholet** [comment]
> Adding `AsMut` and `AsRef` later on would be "allowed breakage", per RFC 1105.

I don’t think this is correct. That RFC says:

> That means that any breakage in a minor release must be very “shallow”: it must always be possible to locally fix the problem through some kind of disambiguation _that could have been done in advance_ (by using more explicit forms) or other annotation (like disabling a lint). 

That doesn’t hold in this case.

**Comment by rust-rfcbot** [comment]
@the8472 has proposed to merge this. The next step is review by the rest of the tagged team members:

* [x] @Amanieu
* [x] @BurntSushi
* [x] @dtolnay
* [x] @joshtriplett
* [x] @the8472

Concerns:

* improve-example (https://github.com/rust-lang/rust/issues/98407#issuecomment-4609027262)
* ~~naming~~ resolved by https://github.com/rust-lang/rust/issues/98407#issuecomment-4609027262
* ~~naming-as-pin-ref~~ resolved by https://github.com/rust-lang/rust/issues/98407#issuecomment-4642050337

Once a majority of reviewers approve (and at most 2 approvals are outstanding), this will enter its final comment period. If you spot a major issue that hasn't been raised at any point in this process, please speak up!

cc @rust-lang/libs-ping: FCP proposed for libs, please feel free to register concerns.
See [this document](https://github.com/rust-lang/rfcbot-rs/blob/master/README.md) for info about what commands tagged team members can give me.

**Comment by joseph-gio** [comment]
The proposed `read` api is safe _only if the purpose of the type is to make things sync_. If the purpose of the type is to enforce exclusive access, then `read` is unsound. if the type is called `Exclusive` but its purpose is not to enforce exclusive access, then that's going to mislead people who might be looking for that property

**Comment by joseph-gio** [comment]
it might make more sense to have two separate types for this.
* `SyncSafe` -- a wrapper type that makes a non-`Sync` type sync. Exposes a `read` method for types that are already sync.
* `Exclusive` -- enforces that the contained value is only accessible with exclusive access. This is orthogonal to the `Sync`-ness of the contained type, and is enforced no matter what.

Not trying to delay stabilization (I couldn't even if I wanted to) -- as a user I just don't want to be stuck with a bad name until the end of time. I feel like a name along the lines of `AlwaysSync` or `SyncSafe` should be uncontroversial

**Comment by pthariensflame** [comment]
The rename to `SyncView` has now been merged (as #153038); the only things we personally are aware of left to do before stabilization are:
- bikeshedding of certain method names (`as_pin` *versus* `as_pin_ref` comes to mind)
- addressing the issue with the documentation example that was pointed out in #146245.
- deciding the fate of the `impl Debug` (currently, the only nontrivial API surface that `&SyncView<impl ?Sync>` unconditionally implements).

Any additional APIs (*e.g.*, our proposed `impl Iterator`) needn’t block stabilization, we would think.

**Comment by BurntSushi** [comment]
* It seems like we should go with `as_pin_ref` since [`Option::as_pin_ref`](https://doc.rust-lang.org/nightly/std/option/enum.Option.html#method.as_pin_ref) exists.
* Getting a better example seems wise. Maybe @alice-i-cecile has one in her back-pocket?
* I think the existing `impl Debug` is probably our only choice here right? It's unfortunate it won't include the value, but constraining the `Debug` impl to `T: Sync` seems like a worse outcome.

@rfcbot resolve naming

@rfcbot concern naming-as-pin-ref

@rfcbot concern improve-example

**Comment by alice-i-cecile** [comment]
My back-pocket has only lint at the moment, but [I'll put the question to the community](https://discord.com/channels/691052431525675048/749335865876021248/1511763616750239944) :)

**Comment by joseph-gio** [comment]
Couldn't `Debug` be specialized based on `T: Sync` ?

**Comment by joseph-gio** [comment]
Also, a classic example for `SyncCell` in bevy is using it to store an `std::sync::mpsc::Receiver<T>` as a `Resource`:

```rust
pub struct Msg { ... }

// Bevy requires any type implementing `Resource` to be `Send + Sync + 'static`.
// `Receiver` is `Send`, but not `Sync`. Thus, it cannot be used in a `Resource` by default.
// By wrapping in `SyncView`, we may store it as a resource, but in order to actually use it we require a `ResMut`
#[derive(Resource)]
pub struct MsgReceiver(SyncView<Receiver<Msg>>);

pub fn my_message_receiver_system(
    // bevy guarantees that the pointee of a `ResMut` will only be accessed by a single thread at a time,
    // which upholds the "single consumer" requirement of `std::sync::mpsc::Receiver`.
    mut receiver: ResMut<MessageReceiver>,
    // note that it would be possible to write `Res<MessageReceiver>` here -- but this would not violate the "single
    // consumer" requirement, since the `SyncView` makes the `Res` useless, and thus harmless.
) {
    // do something with the receiver...
}
```

**Comment by pthariensflame** [comment]
> * I think the existing `impl Debug` is probably our only choice here right? It's unfortunate it won't include the value, but constraining the `Debug` impl to `T: Sync` seems like a worse outcome.

Wild idea: would an `impl<'a, T: Debug> Debug for &'a mut SyncView<T>` work?

**Comment by BurntSushi** [comment]
> Couldn't `Debug` be specialized based on `T: Sync` ?

I'm unsure what our policy is for using specialization for things like this. cc @rust-lang/lang I think?

> > * I think the existing `impl Debug` is probably our only choice here right? It's unfortunate it won't include the value, but constraining the `Debug` impl to `T: Sync` seems like a worse outcome.
> 
> Wild idea: would an `impl<'a, T: Debug> Debug for &'a mut SyncView<T>` work?

This doesn't work if you `derive(Debug)` on a struct containing a `SyncView<T>` though right?

**Comment by BurntSushi** [comment]
> My back-pocket has only lint at the moment, but [I'll put the question to the community](https://discord.com/channels/691052431525675048/749335865876021248/1511763616750239944) :)

If I click this link it says I don't have access. FYI. So if anyone comes up with any, please post them back here!

**Comment by joseph-gio** [comment]
I posted an example above -- a bit bevy-specific though

**Comment by BurntSushi** [comment]
Yeah happy to see something like that turned into a PR if it can be abstracted above a specific crate and discussed in more general/broad terms.

**Comment by BurntSushi** [comment]
@rfcbot resolve naming-as-pin-ref

Ref https://github.com/rust-lang/rust/pull/157521

If someone wants to submit a PR with a good example here, then I think that should unblock the FCP!

**Comment by tisonkun** [comment]
> If someone wants to submit a PR with a good example here, then I think that should unblock the FCP!

@BurntSushi I propose one at https://github.com/rust-lang/rust/pull/162621

Somehow from [this function](https://github.com/apache/asyncband/blob/03a5b588e91c51195da200803caa20dbd402515a/asyncband/src/mpsc/unbounded/receiver.rs#L37-L41) that I believe a stablized `SyncView` would help as well.

**Comment by tisonkun** [comment]
https://github.com/rust-lang/rust/pull/162621 has been merged.

Shall we start the FCP?

cc @BurntSushi @JohnTitor @joshtriplett

**Comment by clarfonthey** [comment]
Note: FCP is currently in progress, only waiting on a better example to be merged.

**Comment by tisonkun** [comment]
> Note: FCP is currently in progress, only waiting on a better example to be merged.

Thanks. I saw https://github.com/rust-lang/rust/issues/98407#issuecomment-3428256024 now.

Maybe we can resolve the concern. If it is desired, I can try to file the stabilization PR.

**Comment by clarfonthey** [comment]
Huh, yeah, I guess we did update the example.

@BurntSushi when you have time, do you think the new example is sufficient? https://doc.rust-lang.org/nightly/std/sync/struct.SyncView.html

**Comment by BurntSushi** [comment]
@rfcbot resolve improve-example

**Comment by clarfonthey** [comment]
@BurntSushi apparently the command is resolved, not resolve: https://github.com/rust-lang/rfcbot-rs#concerns

🙃

**Comment by BurntSushi** [comment]
@rfcbot resolved improve-example

**Comment by tgross35** [comment]
I thought they both work, but it seems like that one didn't resolve it either? Wonder if it's related to FCP team membership updates.

**Comment by clarfonthey** [comment]
Shouldn't be…

**Comment by clarfonthey** [comment]
Okay, actually, the docs are wrong, and it worked earlier in this thread? I think rfcbot is just broken.

**Comment by ChrisDenton** [comment]
@rfcbot fcp resolve improve-example

**Comment by tgross35** [comment]
Guess those examples just aren't improved enough 🙃 

On a more serious note, I assume https://github.com/rust-lang/rust/issues/146245 can be closed?
